src/rulesets/oas2/functions/__tests__/oas2UnusedDefinition.test.ts

```diff
@@ -0,0 +1,101 @@
+import { getLocationForJsonPath, parseWithPointers } from '@stoplight/json';
+import { DiagnosticSeverity } from '@stoplight/types';
+import { RuleType, Spectral } from '../../../../index';
+import { rules as oas2Rules } from '../../../oas2/index.json';
+import oas2UnusedDefinition from '../oas2UnusedDefinition';
+
+describe('oas2UnusedDefinition', () => {
```

Comment:
@P0lip @philsturgeon are there some corner cases tests you'd like me to add here to ensure the rule behaves as expected?

Reply:
There are at least 2 corner cases I can think of at the moment.

- "remote-local" like reference:

```json
// foo.json
{
 "swagger": "2.0",
"x-hook": {
    "$ref": "./foo.json#/definitions/Hooked"
 },
 "definitions": {
    "Hooked": {
       "type": "object"
     }
  }
}
```

- circular reference case

```json
// foo.json
{
   "swagger": "2.0",
   "x-hook": {
    "$ref": "./foo.json#/definitions/Hooked"
   },
   "definitions": {
      "Hooked": {
         "type": "object"
      }
    }
}
```
```json
// bar.json
{
  "foo": {
    "$ref": "./foo.json#/definitions/Hooked"
  }
}
```

Reply:
I suspect these two might fail, as we might have insufficient information.
Happy to help with solving that if it happens to be an issue.

Reply:
@P0lip Awesome. Let me see what I can come up with with this input. I fear I will indeed need some help. I'll ping you when I'm crying too much.

Reply:
@nulltoken
`Resolved` contains `refMap` you might need to use to retrieve collected feedback on $refs.

`lintNode` receives `Resolved`, but neither does custom function. You can simply pass it down, though.
Just add it to `IFunctionValues`, i.e.
```ts
export interface IFunctionValues {
  original: any;
  given: any;
  resolved: Resolved;
}
```
and simply pass `resolved` to all calls :)

Reply:
Starting to work on this now that Karma is all happy...

Reply:
Great, thank you so much.
Do not hesitate to reach out if you hit a wall.

Reply:
@P0lip So the `Resolved` thingie was a very interesting hint. Thank you! That led me to completely change the approach. Going that way, I might not need the iterator any more.

What do you think so far. Am I going the wrong way?

I haven't worked on the additional corner case tests yet, but that's the next item on my list.

Reply:
> Going that way, I might not need the iterator any more.
> What do you think so far. Am I going the wrong way?

No, not at all. The direction sounds totally right. The iteration might be in fact not needed anymore, since `Resolved` should have enough information to list unused definitions.

Reply:
@P0lip So I think I tamed the `remote-local` like case (in the latest commit).

However, I 'm not sure to understand the circular reference case. Is there a typo somewhere in what you want me test or am I overlooking the obvious edge case. That doesn't seem very circular to me ;-)

Reply:
@P0lip So I went ahead and tried to work on that circular corner case.

I've started small. Only half a circle.

And I hit a wall.

We load File A. File A holds a definition. File A also references File B which, in turn, references back the definition of File A.

So File A is indirectly self-referencing itself.

It gets properly resolved (Yeah!). However, the refMap only seems to only contain refs related to A.
I seem to find traces of the File B refs (in `parsed`) but not in a state that seems immediately actionable to me.

Of course, my unit test amazingly. Tada!

Any idea? _(Please say "Yes")_

Reply:
> somewhere in what you want me test or am I overlooking the obvious edge case. That doesn't seem very circular to me ;-)

Ah sorry, I meant `bar.json` here
```
// foo.json
{
   "swagger": "2.0",
   "x-hook": {
    "$ref": "./bar.json#/foo"
   },
```
I hope you've figured that part out.


> It gets properly resolved (Yeah!). However, the refMap only seems to only contain refs related to A.

You are correct, `refMap` does not contain any information about $refs met in other referenced files. It collects feedback only for file you execute the dereferencing process on, *however*, Spectral still gathers certain stuff on its own, so that we can map errors with proper files.
You should be able to consume that value.

Take a look at the [_processExternalRef](https://github.com/stoplightio/spectral/blob/6a6d477674efff3d73f3057bc9740a3e61f4eba4/src/spectral.ts#L219) method to see what's available.
`_parsedMap` might contain everything you need.
It's exposed on `Resolved.

LMK if it does the trick. 
Once again, thank you for all the effort. Really appreciated!

Reply:
If `parsedMap` has insufficient information, feel free to include anything that's needed.
Let me know how it goes

Reply:
One extra possibility. Recently we've added graph dependency to json-ref-resolver (https://github.com/stoplightio/json-ref-resolver/pull/104). We are still on the older version, but feel free to bump it. `_parsedMap` is likely to be eventually removed in favor of that dependency graph.

Reply:
> > somewhere in what you want me test or am I overlooking the obvious edge case. That doesn't seem very circular to me ;-)
> 
> Ah sorry, I meant `bar.json` here

This use case should be covered through the "indirectly self-referencing" test case, then.